framework/src/modules/token/api.ts

```diff
@@ -52,55 +63,87 @@ export class TokenAPI extends BaseAPI {
 		apiContext: APIContext,
 		senderAddress: Buffer,
 		recipientAddress: Buffer,
-		_tokenID: TokenID,
+		tokenID: TokenID,
 		amount: bigint,
 	): Promise<void> {
+		const canonicalTokenID = await this._getCanonicalTokenID(apiContext, tokenID);
 		const userStore = apiContext.getStore(this.moduleID, STORE_PREFIX_USER);
-		const sender = await userStore.getWithSchema<UserStoreData>(senderAddress, userStoreSchema);
-		if (sender.availableBalance < amount + this._minBalance) {
+		const sender = await userStore.getWithSchema<UserStoreData>(
+			getUserStoreKey(senderAddress, canonicalTokenID),
+			userStoreSchema,
+		);
+		if (sender.availableBalance < amount) {
 			throw new Error(
 				`Sender ${senderAddress.toString(
 					'hex',
-				)} balance ${sender.availableBalance.toString()} is not sufficient for ${(
-					amount + this._minBalance
-				).toString()}`,
+				)} balance ${sender.availableBalance.toString()} is not sufficient for ${amount.toString()}`,
 			);
 		}
+
 		sender.availableBalance -= amount;
-		await userStore.setWithSchema(senderAddress, sender, userStoreSchema);
+
+		await userStore.setWithSchema(
+			getUserStoreKey(senderAddress, canonicalTokenID),
+			sender,
+			userStoreSchema,
+		);
 
 		let recipient: UserStoreData;
+		let recipientExist = false;
 		try {
-			recipient = await userStore.getWithSchema<UserStoreData>(recipientAddress, userStoreSchema);
+			recipient = await userStore.getWithSchema<UserStoreData>(
+				getUserStoreKey(recipientAddress, canonicalTokenID),
+				userStoreSchema,
+			);
+			recipientExist = true;
```

Comment:
I think I misunderstood the concept. Maybe it would be better to avoid the dependency to auth module, and rather only rely on `recipient holds any token`?
Also, user data is never deleted, so if there is a data corresponding to the address, I think it means `nonce > 0` indirectly

Reply:
"User data is never deleted", ok, I guess that's fine. I thought that we would remove entries if the amount of tokens goes to 0. This way, we can clean up the store of all those empty accounts.

But if you tell me that 0 token entries are not really an issue (and probably they are not), we could leave them. We would need to update the LIP though.

How bad is it to rely on nonce?

Reply:
If we can resolve the issue i would try not to add dependency, but it's also not so bad i guess.

how about we still remove the entry if becomes empty, but if all becomes empty, user need to pay min balance to re-initialize the account?

Reply:
I think we settled on not removing empty entries, and the min balance logic is now quite different. Will resolve this comment.